### PR TITLE
Convert selected chat conversation state into routeTree route

### PR DESCRIPTION
**Pull request by chromakode**
Working towards getting attachment popups as routes, I needed to be able to express the chat you're on as a route.

Chat routes look like this: `/tab:chatTab/<conversationId>`

It's important to create a route for each conversation. If we store anything in `routeState`, we'll want separate state for each tab of the conversation list. Thus, `routeTree` needed to be extended so that the children of a route can be a function. Tests included!

The rest of this is just moving state around. There's now a `nothingSelected` sentinel value to represent no tab selected yet. Deriving the selected conversation out of the route state is a little terse, but I'm not sure what this code is doing is worth generalizing into a helper.

:eyeglasses: @keybase/react-hackers

### CODE REVIEW

File: shared/actions/chat.js [hunk 1]
```diff
@@ -72,7 +73,10 @@ const {
 
 const {conversationIDToKey, keyToConversationID, InboxStateRecord, MetaDataRecord, makeSnippet, serverMessageToMessageBody} = Constants
 
-const _selectedSelector = (state: TypedState) => state.chat.get('selectedConversation')
+const _selectedSelector = (state: TypedState) => {
+  const selected = getPath(state.routeTree.routeState, [chatTab]).last()
```
Comment (chrisnojima) [inline shared/actions/chat.js]: per our discussion this make assumptions about the route that'll likely change (esp. w/ modals)

Comment (chromakode) [inline shared/actions/chat.js]: @chrisnojima Fixed and merge conflicts resolved. @keybase/react-hackers ready for review!

**Comment by chromakode** [comment]
:eyeglasses: @keybase/react-hackers bump

File: shared/route-tree/index.js [hunk 2]
```diff
@@ -28,7 +28,7 @@ type RouteDefParams<P> = {
   defaultSelected?: string,
   tags?: LeafTags,
   initialState?: {},
-  children: {},
+  children: {} | (name: string) => RouteDefNode,
```
Comment (MarcoPolo) [inline shared/route-tree/index.js]: is children type of `{[key: string]: RouteDefNode}` normally?

File: shared/login/login/routes.js [hunk 3]
```diff
@@ -44,9 +44,9 @@ const recursiveLazyRoutes = I.Seq({
     component: Success,
   },
 })
-  .map(routeData => () => new RouteDefNode({
+  .map(routeData => new RouteDefNode({
     ...routeData,
-    children: recursiveLazyRoutes,
+    children: name => recursiveLazyRoutes.get(name),
```
Comment (MarcoPolo) [inline shared/login/login/routes.js]: why this change?

File: shared/actions/chat.js [hunk 4]
```diff
@@ -916,12 +932,11 @@ function * _updateReachability (action: UpdateReachability): SagaGenerator<any,
 
 function * _sendNotifications (action: AppendMessages): SagaGenerator<any, any> {
   const appFocused = yield select(_focusedSelector)
-  const conversationIDKey = yield select(_selectedSelector)
   const selectedTab = yield select(_routeSelector)
   const chatTabSelected = (selectedTab === chatTab)
 
   // Only send if you're not looking at it
-  if (conversationIDKey !== action.payload.conversationIDKey || !appFocused || !chatTabSelected) {
+  if (action.isSelected || !appFocused || !chatTabSelected) {
```
Comment (chromakode) [inline shared/actions/chat.js]: Oops! Good catch, thanks!

File: shared/route-tree/index.js (hunk 2 above)
Comment (chromakode) [inline shared/route-tree/index.js]: It's actually `{[key: string]: RouteDefParams<P> | () => RouteDefNode} | (name: string) => RouteDefNode`. Updated!
